src/java.base/share/classes/java/util/DualPivotQuicksort.java

```diff
@@ -120,6 +123,78 @@ private DualPivotQuicksort() {}
      */
     private static final int MAX_RECURSION_DEPTH = 64 * DELTA;
 
+    /**
+     * Represents a function that accepts the array and sorts the specified range
+     * of the array into ascending order.
+     */
+    @FunctionalInterface
+    private static interface SortOperation<A> {
+        /**
+         * Sorts the specified range of the array.
+         *
+         * @param a the array to be sorted
+         * @param low the index of the first element, inclusive, to be sorted
+         * @param high the index of the last element, exclusive, to be sorted
+         */
+        void sort(A a, int low, int high);
+    }
+
+    /**
+     * Sorts the specified range of the array into ascending numerical order.
+     *
+     * @param elemType the class of the elements of the array to be sorted
+     * @param array the array to be sorted
+     * @param offset the relative offset, in bytes, from the base address of
+     * the array to sort, otherwise if the array is {@code null},an absolute
+     * address pointing to the first element to sort from.
+     * @param low the index of the first element, inclusive, to be sorted
+     * @param high the index of the last element, exclusive, to be sorted
+     * @param so the method reference for the fallback implementation
+     */
+    @IntrinsicCandidate
+    @ForceInline
+    private static <A> void sort(Class<?> elemType, A array, long offset, int low, int high, SortOperation<A> so) {
+        so.sort(array, low, high);
```

Comment:
I'm late to the party, but how does the fallback implementation work (or is intended to) for off-heap case (null + absolute base address)?

Reply:
Also, for on-heap case the fallback implementation is equivalent to intrinsified case only when offset points at the 0th element of the array.

Reply:
@iwanowww Yes, you are late to the party :). The fallback implementation could be similar to the vectorizedMismatch regarding base/offset for non heap case. Please see java.base/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java.
Yes, the fallback implementation for non intrinsic case is kept to be equivalent to what was before the SIMD sort PR. This is done to not affect the fallback performance on other platforms.

Reply:
The problem with fallback implementation as it is now is that it doesn't take into account the offset. It's completely broken for off-heap case and works for on-heap case only because the implementation always passes primitive array base offset.